compiler/analyzer.ml

```diff
@@ -283,14 +304,9 @@ and check_binop env e1 op e2 =
         env', Sast.Expr(Sast.Binop(ew1', op, ew2'), result_type)
       else binop_error typ1 op typ2
 
-    (* Equality operations - overloaded, no constraining can be done *)
-    | Eq | Neq -> 
-      let is_valid_equality = function
-        | Num | Bool | String | Unconst -> true
-        | _ -> false in 
-      if is_valid_equality typ1 && is_valid_equality typ2 then 
-        env', Sast.Expr(Sast.Binop(ew1, op, ew2), Bool)
-      else binop_error typ1 op typ2
+    (* Equality operations - overloaded, no constraining can be done, can take
```

Comment:
I take back what I said about lists, you can compare those in Python with `==`. But if we allow all types then what happens if we do

```
do (comp = (x, y) -> return x == y) == (f = (x) -> return 2)
```

I think that will be legal but now we need to move the function declarations above the equality in the PAST, as we're doing for anonymous functions.